### PR TITLE
[3758] Store all form payload in a context

### LINKED ISSUE #3758: Store form and property payload in a context

### PR DESCRIPTION
**Pull request by mcharfadi**
Bug: https://github.com/eclipse-sirius/sirius-web/issues/3758

# Pull request template

## General purpose
What is the main goal of this pull request?
- [ ] Bug fixes
- [ ] New features
- [ ] Documentation
- [ ] Cleanup
- [ ] Tests
- [ ] Build / releng

## Project management
- [ ] Has the pull request been added to the relevant project and milestone? (Only if you know that your work is part of a specific iteration such as the current one)
- [ ] Have the `priority:` and `pr:` labels been added to the pull request? (In case of doubt, start with the labels `priority: low` and `pr: to review later`)
- [ ] Have the relevant issues been added to the pull request?
- [ ] Have the relevant labels been added to the issues? (`area:`, `difficulty:`, `type:`)
- [ ] Have the relevant issues been added to the same project and milestone as the pull request?
- [ ] Has the `CHANGELOG.adoc` been updated to reference the relevant issues?
- [ ] Have the relevant API breaks been described in the `CHANGELOG.adoc`? (Including changes in the GraphQL API)
- [ ] In case of a change with a visual impact, are there any screenshots in the `CHANGELOG.adoc`? For example in `doc/screenshots/2022.5.0-my-new-feature.png`

## Architectural decision records (ADR)
- [ ] Does the title of the commit contributing the ADR start with `[doc]`?
- [ ] Are the ADRs mentioned in the relevant section of the  `CHANGELOG.adoc`?

## Dependencies
- [ ] Are the new / upgraded dependencies mentioned in the relevant section of the `CHANGELOG.adoc`?
- [ ] Are the new dependencies justified in the `CHANGELOG.adoc`?

## Frontend

This section is not relevant if your contribution does not come with changes to the frontend.

### General purpose
- [ ] Is the code properly tested? (Plain old JavaScript tests for business code and tests based on React Testing Library for the components)

### Typing
We need to improve the typing of our code, as such, we require every contribution to come with proper TypeScript typing for both changes contributing new files and those modifying existing files.
Please ensure that the following statements are true for each file created or modified (this may require you to improve code outside of your contribution).

- [ ] Variables have a proper type
- [ ] Functions’ arguments have a proper type
- [ ] Functions’ return type are specified
- Hooks are properly typed:
	- [ ] `useMutation<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useQuery<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useSubscription<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useMachine<CONTEXT_TYPE, EVENTS_TYPE>(…)`
	- [ ] `useState<STATE_TYPE>(…)`
- [ ] All components have a proper typing for their props
- [ ] No useless optional chaining with `?.` (if the GraphQL API specifies that a field cannot be `null`, do not treat it has potentially `null` for example)
- [ ] Nullable values have a proper type (for example `let diagram: Diagram | null = null;`)

## Backend

This section is not relevant if your contribution does not come with changes to the backend.

### General purpose
- [ ] Are all the event handlers tested?
- [ ] Are the event processor tested?
- [ ] Is the business code (services) tested?
- [ ] Are diagram layout changes tested?

### Architecture
- [ ] Are data structure classes properly separated from behavioral classes?
- [ ] Are all the relevant fields final?
- [ ] Is any data structure mutable? If so, please write a comment indicating why
- [ ] Are behavioral classes either stateless or side effect free?

### How to test this PR?
_Please describe here the various use cases to test this pull request_

- [ ] Has the Kiwi TCMS test suite been updated with tests for this contribution?

### CODE REVIEW

File: packages/forms/frontend/sirius-components-forms/src/representations/FormRepresentation.tsx [hunk 1]
```diff
@@ -76,120 +51,70 @@ const useFormRepresentationStyles = makeStyles()((theme) => ({
...
-              <div className={classes.toolbarAction} key={toolbarAction.id}>
-                <ToolbarAction
-                  editingContextId={editingContextId}
-                  formId={form.id}
-                  readOnly={readOnly}
-                  widget={toolbarAction}
-                />
-              </div>
-            ))}
+  const renderedForm = useMemo(() => {
```
Comment (sbegaudeau) [inline packages/forms/frontend/sirius-components-forms/src/representations/FormRepresentation.tsx]: Please stop with the `useMemo` `memo`, `useCallback`etc unless we have a performance issue identified that needs to be fixed (i.e. a Github issue). Premature optimization does not help and React 19 is supposed to provide that for free so we will have to remove this kind of code quite soon.

File: packages/forms/frontend/sirius-components-forms/src/representations/useFormSubscription.tsx [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+    const { data: gqlFormEventSubscription } = data;
+    if (gqlFormEventSubscription) {
+      const { formEvent: payload } = gqlFormEventSubscription;
+      setState((prevState) => ({ ...prevState, payload }));
+    }
+  };
+
+  const onComplete = () => setState((prevState) => ({ ...prevState, complete: true }));
+
+  const { error, loading } = useSubscription<GQLFormEventSubscription, GQLFormEventVariables>(
```
Comment (sbegaudeau) [inline packages/forms/frontend/sirius-components-forms/src/representations/useFormSubscription.tsx]: Why not destructure data here like that `{ data, error, loading }` and remove the payload from the state? You would also remove one re-rendering with the setState line57 which would disappear.

**Review by sbegaudeau — CHANGES_REQUESTED** [review]
Could you also rebase the PR to merge it quickly after that, thanks

File: packages/forms/frontend/sirius-components-forms/src/representations/useFormSubscription.tsx (hunk 2 above)
Comment (mcharfadi) [reply packages/forms/frontend/sirius-components-forms/src/representations/useFormSubscription.tsx]: Okay I modified all the custom hook to act like this.

File: packages/forms/frontend/sirius-components-forms/src/representations/FormRepresentation.tsx (hunk 1 above)
Comment (mcharfadi) [reply packages/forms/frontend/sirius-components-forms/src/representations/FormRepresentation.tsx]: Seemed worth it to me there, but I also took the wrong habit to use these too much.
I'll remove it there.

File: packages/forms/frontend/sirius-components-forms/src/representations/FormRepresentation.tsx [hunk 3]
```diff
@@ -76,83 +51,37 @@ const useFormRepresentationStyles = makeStyles()((theme) => ({
...
-  const { id, formId, form, message } = context;
   const [state, setState] = useState<FormRepresentationState>({
-    payload: null,
+    form: null,
   });
 
-  /**
-   * Displays an other form if the selection indicates that we should display another properties view.
-   */
+  const { payload } = useFormSubscription(editingContextId, representationId);
```
Comment (sbegaudeau) [inline packages/forms/frontend/sirius-components-forms/src/representations/FormRepresentation.tsx]: You are not using the complete information to indicate that the form has been deleted like it was before

File: packages/sirius-web/frontend/sirius-web-application/src/diagrams/useDiagramFilterSubscription.tsx [hunk 4]
```diff
@@ -59,20 +60,19 @@ export const useDiagramFilterSubscription = (
...
+  const onComplete = () => setState((prevState) => ({ ...prevState, complete: true }));
+
   const onData = ({ data }: OnDataOptions<GQLDiagramFilterEventSubscription>) => {
     const { data: gqlDiagramFilterEventSubscription } = data;
     if (gqlDiagramFilterEventSubscription) {
       const { diagramFilterEvent: payload } = gqlDiagramFilterEventSubscription;
       if (isFormRefreshedEventPayload(payload)) {
-        const { form } = payload;
-        setState((prevState) => ({ ...prevState, form, complete: false }));
+        setState((prevState) => ({ ...prevState, complete: false }));
```
Comment (sbegaudeau) [inline packages/sirius-web/frontend/sirius-web-application/src/diagrams/useDiagramFilterSubscription.tsx]: You can remove the rest of the code from this onData, we don't anything other than "if we receive some data, then it's not complete anymore", we don't need to do that only on a form refreshed event payload

File: packages/sirius-web/frontend/sirius-web-application/src/diagrams/useDiagramFilterSubscription.tsx [hunk 5]
```diff
@@ -92,7 +92,7 @@ export const useDiagramFilterSubscription = (
 
   return {
     loading,
-    form: state.form,
+    payload: !!data?.diagramFilterEvent ? data.diagramFilterEvent : null,
```
Comment (sbegaudeau) [inline packages/sirius-web/frontend/sirius-web-application/src/diagrams/useDiagramFilterSubscription.tsx]: Why not `data?.diagramFilterEvent ?? null`?

File: packages/sirius-web/frontend/sirius-web-application/src/views/edit-project/workbench-views/useDetailsViewSubscription.tsx [hunk 6]
```diff
@@ -60,21 +61,20 @@ export const useDetailsViewSubscription = (
...
+
   const onData = ({ data }: OnDataOptions<GQLDetailsEventSubscription>) => {
     const { data: gqlDetailsEventSubscription } = data;
     if (gqlDetailsEventSubscription) {
       const { detailsEvent: payload } = gqlDetailsEventSubscription;
       setState((prevState) => ({ ...prevState, payload, complete: false }));
       if (isFormRefreshedEventPayload(payload)) {
-        const { form } = payload;
-        setState((prevState) => ({ ...prevState, form, complete: false }));
+        setState((prevState) => ({ ...prevState, complete: false }));
```
Comment (sbegaudeau) [inline packages/sirius-web/frontend/sirius-web-application/src/views/edit-project/workbench-views/useDetailsViewSubscription.tsx]: Same comment as the previous hook, you don't need any of those "if" to perform the `setState`

File: packages/sirius-web/frontend/sirius-web-application/src/views/edit-project/workbench-views/useDetailsViewSubscription.tsx [hunk 7]
```diff
@@ -94,8 +94,7 @@ export const useDetailsViewSubscription = (
 
   return {
     loading,
-    form: state.form,
-    payload: state.payload,
+    payload: !!data?.detailsEvent ? data.detailsEvent : null,
```
Comment (sbegaudeau) [inline packages/sirius-web/frontend/sirius-web-application/src/views/edit-project/workbench-views/useDetailsViewSubscription.tsx]: Can be simplified

File: packages/sirius-web/frontend/sirius-web-application/src/views/edit-project/workbench-views/useRelatedElementsViewSubscription.tsx [hunk 8]
```diff
@@ -60,29 +61,28 @@ export const useRelatedElementsViewSubscription = (
...
+  const onComplete = () => setState((prevState) => ({ ...prevState, complete: true }));
+
   const onData = ({ data }: OnDataOptions<GQLRelatedElementsEventSubscription>) => {
     const { data: gqlRelatedElementsEventSubscription } = data;
     if (gqlRelatedElementsEventSubscription) {
       const { relatedElementsEvent: payload } = gqlRelatedElementsEventSubscription;
       if (isFormRefreshedEventPayload(payload)) {
-        const { form } = payload;
-        setState((prevState) => ({ ...prevState, form, complete: false }));
+        setState((prevState) => ({ ...prevState, complete: false }));
```
Comment (sbegaudeau) [inline packages/sirius-web/frontend/sirius-web-application/src/views/edit-project/workbench-views/useRelatedElementsViewSubscription.tsx]: Same comment

File: packages/sirius-web/frontend/sirius-web-application/src/views/edit-project/workbench-views/useRelatedElementsViewSubscription.tsx [hunk 9]
```diff
@@ -93,7 +93,7 @@ export const useRelatedElementsViewSubscription = (
 
   return {
     loading,
-    form: state.form,
+    payload: !!data?.relatedElementsEvent ? data.relatedElementsEvent : null,
```
Comment (sbegaudeau) [inline packages/sirius-web/frontend/sirius-web-application/src/views/edit-project/workbench-views/useRelatedElementsViewSubscription.tsx]: Same comment too

File: packages/sirius-web/frontend/sirius-web-application/src/views/edit-project/workbench-views/useRepresentationsViewSubscription.tsx [hunk 10]
```diff
@@ -60,29 +62,28 @@ export const useRepresentationsViewSubscription = (
...
+  const onComplete = () => setState((prevState) => ({ ...prevState, complete: true }));
+
   const onData = ({ data }: OnDataOptions<GQLRepresentationsEventSubscription>) => {
     const { data: gqlRepresentationsEventSubscription } = data;
     if (gqlRepresentationsEventSubscription) {
       const { representationsEvent: payload } = gqlRepresentationsEventSubscription;
       if (isFormRefreshedEventPayload(payload)) {
-        const { form } = payload;
-        setState((prevState) => ({ ...prevState, form, complete: false }));
+        setState((prevState) => ({ ...prevState, complete: false }));
```
Comment (sbegaudeau) [inline packages/sirius-web/frontend/sirius-web-application/src/views/edit-project/workbench-views/useRepresentationsViewSubscription.tsx]: Same comment

File: packages/sirius-web/frontend/sirius-web-application/src/views/edit-project/workbench-views/useRepresentationsViewSubscription.tsx [hunk 11]
```diff
@@ -93,7 +94,7 @@ export const useRepresentationsViewSubscription = (
 
   return {
     loading,
-    form: state.form,
+    payload: !!data?.representationsEvent ? data.representationsEvent : null,
```
Comment (sbegaudeau) [inline packages/sirius-web/frontend/sirius-web-application/src/views/edit-project/workbench-views/useRepresentationsViewSubscription.tsx]: Same comment

**Review by sbegaudeau — CHANGES_REQUESTED** [review]
I'll take into account the various comments that I've added to this PR and ensure that it gets merge to release the Sirius Web RC1.

File: packages/forms/frontend/sirius-components-forms/src/representations/FormRepresentation.tsx (hunk 3 above)
Comment (sbegaudeau) [reply packages/forms/frontend/sirius-components-forms/src/representations/FormRepresentation.tsx]: Done

File: packages/sirius-web/frontend/sirius-web-application/src/diagrams/useDiagramFilterSubscription.tsx (hunk 5 above)
Comment (sbegaudeau) [reply packages/sirius-web/frontend/sirius-web-application/src/diagrams/useDiagramFilterSubscription.tsx]: Done

**Comment by sbegaudeau** [comment]
I'll wait for the build, merge it and release the Sirius Web 2024.9.0 RC1 (which will be the milestone 2024.7.11)

**Review by sbegaudeau — APPROVED** [review]

**Pull request merged**